config/envoyconfig/listeners.go

```diff
@@ -518,65 +523,46 @@ func (b *Builder) buildRouteConfiguration(name string, virtualHosts []*envoy_con
 	}, nil
 }
 
-func (b *Builder) buildDownstreamTLSContext(ctx context.Context,
+func (b *Builder) buildDownstreamTLSContextMulti(
+	ctx context.Context,
 	cfg *config.Config,
-	serverName string,
-) *envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext {
-	cert, err := cfg.GetCertificateForServerName(serverName)
-	if err != nil {
-		log.Warn(ctx).Str("domain", serverName).Err(err).Msg("failed to get certificate for domain")
-		return nil
-	}
-
-	err = validateCertificate(cert)
-	if err != nil {
-		log.Warn(ctx).Str("domain", serverName).Err(err).Msg("invalid certificate for domain")
-		return nil
+	certs []tls.Certificate) (
+	*envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext,
+	error,
+) {
+	envoyCerts := make([]*envoy_extensions_transport_sockets_tls_v3.TlsCertificate, 0, len(certs))
+	for i := range certs {
+		cert := &certs[i]
+		if err := validateCertificate(cert); err != nil {
+			return nil, fmt.Errorf("invalid certificate for domain %s: %w", cert.Leaf.Subject.CommonName, err)
+		}
+		envoyCert := b.envoyTLSCertificateFromGoTLSCertificate(ctx, cert)
+		envoyCerts = append(envoyCerts, envoyCert)
 	}
+	return &envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext{
+		CommonTlsContext: &envoy_extensions_transport_sockets_tls_v3.CommonTlsContext{
+			TlsParams:             tlsParams,
+			TlsCertificates:       envoyCerts,
+			AlpnProtocols:         getALPNProtos(cfg.Options),
+			ValidationContextType: b.buildDownstreamValidationContext(ctx, cfg),
+		}}, nil
+}
 
-	var alpnProtocols []string
-	switch cfg.Options.GetCodecType() {
+func getALPNProtos(opts *config.Options) []string {
+	switch opts.GetCodecType() {
 	case config.CodecTypeHTTP1:
-		alpnProtocols = []string{"http/1.1"}
+		return []string{"http/1.1"}
 	case config.CodecTypeHTTP2:
-		alpnProtocols = []string{"h2"}
+		return []string{"h2"}
 	default:
-		alpnProtocols = []string{"h2", "http/1.1"}
-	}
-
-	envoyCert := b.envoyTLSCertificateFromGoTLSCertificate(ctx, cert)
-	return &envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext{
-		CommonTlsContext: &envoy_extensions_transport_sockets_tls_v3.CommonTlsContext{
-			TlsParams:             tlsParams,
-			TlsCertificates:       []*envoy_extensions_transport_sockets_tls_v3.TlsCertificate{envoyCert},
-			AlpnProtocols:         alpnProtocols,
-			ValidationContextType: b.buildDownstreamValidationContext(ctx, cfg, serverName),
-		},
+		return []string{"h2", "http/1.1"}
 	}
 }
 
-func (b *Builder) buildDownstreamValidationContext(ctx context.Context,
+func (b *Builder) buildDownstreamValidationContext(
+	ctx context.Context,
```

Comment:
do the changes to this method mean that if any route uses client certs, all routes will use client certs?

Reply:
I think so yes, if there’s a client cert then it would be provided for any (reused) HTTP/2 connection. however the check of certs happens in authorize, so I suppose they’ll just be ignored for routes that do not require client cert?

Reply:
I believe this would be a change in behavior but it would require testing.